### PR TITLE
[Arm64] Use half barriers for IL Volatile implementation

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

**Comment by sdmaclea** [comment]
@RussKeldorph @Briansull 
cc @dotnet/arm64-contrib @dotnet/jit-contrib 

This is the first step.

Passed GC Stress testing w/o regression

**Comment by sdmaclea** [comment]
First step of #11545

**Comment by briansull** [comment]
Is it possible to have a perf test that measures the relative costs of the different approaches?

File: src/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -2565,7 +2565,7 @@ void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* initBlkNode)
 
     if (initBlkNode->gtFlags & GTF_BLK_VOLATILE)
     {
-        // issue a full memory barrier before volatile an initBlockUnroll operation
+        // issue a full memory barrier before a volatile an initBlockUnroll operation
```
Comment (CarolEidt) [inline src/jit/codegenarm64.cpp]: Typo: remove the "an".

**Comment by jashook** [comment]
@dotnet-bot test Windows_NT Arm64 Checked

**Comment by sdmaclea** [comment]
This appears to be causing intermittent failures in GC stress testing.  I have run GC stress several times and I am seeing one or two failures with this patch that I did not see before this patch (or variants of this patch).

Perhaps .NET Acquire semantics implied by the IL Volatile prefix is not functionally equivalent to Arm64 LoadAcquire?

It is possible the tests are not using the volatile keyword appropriately, but

**Comment by sdmaclea** [comment]
Perhaps the BLK operations should not be relaxed.  

The Volatile prefix on BLK operations is confusing to me anyway.  In the case of a copy, is the Volatile on the Load, Store, or both (must assume both).  It is not required to be atomic.  So does the sync requirement apply to the first or last byte or all the bytes....  

If you assume that the last write could be a ready signal, there should be a store barrier immediately before it.  You could make the same assumption for any byte within the object.  Inserting a barrier between every byte in a BLK copy is definite overkill (and a performance killer).  But it demonstrates why leaving the final barrier as a full barrier seems logical.

**Comment by sdmaclea** [comment]
@russellhadley fyi

**Comment by russellhadley** [comment]
@sdmaclea I think you meant @RussKeldorph :)

**Comment by sdmaclea** [comment]
@russellhadley You are correct...  Thanks.  Note to self ... Read the whole word.  Do not click comment until you review what you wrote....

**Comment by RussKeldorph** [comment]
@sdmaclea Agree this is not obvious.  Agree we don't want barriers between every pair of accesses.  Also agree that `cpblk` needs *full* barriers at at both the "beginning" and the "end."  The question in my mind is whether those barriers should be placed completely outside the memory accesses as they are now or whether the first and last reads should be treated as an acquire and the first and last writes treated as a release.  For the latter option, presumably the half barriers would be collapsed into full barriers between the first read and first write and last read and last write.

`initblk` seems slightly more straightforward, though I think you could make the argument that, in addition to a release barrier before the first write, you should insert a release barrier before the final write as well.

It kinda feels like the difference between the two types of barrier placement is whether information is communicated through the blk operation once *any* bit is accessed or only once *all* bits are accessed.  I'm not sure which is more useful or interesting, and I could believe that it depends on whether you are using it for acquire or release.  Might be worth trying to implement lock-free algorithms with blk ops as semaphores and see what they would look like or if it's even possible without atomicity--assuming someone hasn't already figured this out and written about it.

I'm inclined to restore full barriers to blk ops the way ARM32 does it now and move on, since, again, I'm not an expert. :)

**Comment by sdmaclea** [comment]
@RussKeldorph 

The problem I am facing is that while I had assumed the GC stress issues were fully resolved by the patch copying the arm32 barrier code into arm64, it was only partially true.  

I have been trying to test this patch and a few other barrier related patches, and I always see 1-2 failures which were not present in the tip.  So I ran a recent tip `bad6729` 10 times through GC stress.  In the 10 runs, I found 7 unique GC Stress failures.  So I do not think the tip is completely right.  I am having a hard time relaxing the barriers on the broken tip when it is not fully hardened.

For the block operations, I think we may have it wrong.  I do not think it is reasonable to use a non atomic for implement lock-free algorithms.  I had been assuming it was just a mistake to include block operations in the IL ECMA spec.  I am beginning to think it is required to initialize `static` objects.  Maybe also boxed natives which could(?) be atomic and used for lock free implementations.  

If the Volatile prefix was used for static initiailization, the InitBlk code would needs a Full barrier after initialization.  This is not present in the tip `bad6729` for arm32 or arm64.  It seems a bit unnatural to mark the object as volatile when in reality the reference is volatile.  So this may be a stretch.

For boxed natives, we are not currently implementing the semantics correctly.  We may need to handle small objects differently.

For some of the Block cases, I think I might have missed the important case for object which contain references.  I had skipped them because they were already calling JIT_WriteBarrier to write their references.  But that may not be sufficient.  The JIT_WriteBarrier code is not necessarily using the correct barrier either.

I am also not ready to relax the barriers until I can get a stable baseline.  I will probably go the other way, and be more conservative with the barriers so that I can make this thing stable.

**Comment by BruceForstall** [comment]
@sdmaclea Should this be marked "[WIP]" / tagged "* NO MERGE *" until you are confident in the GC stress results?

**Comment by sdmaclea** [comment]
@dotnet-bot test Tizen armel Cross Debug Build please

**Pull request merged**